pytest-selenium/pages/osweb.py

```diff
@@ -1,14 +1,112 @@
 from pages.base import Page
 from selenium.webdriver.common.by import By
+from selenium.webdriver.support import expected_conditions as expected
+from utils.utility import Utilities
 
 
 class WebBase(Page):
+    URL_TEMPLATE = "/details/books/{book_slug}"
     _root_locator = (By.CSS_SELECTOR, "body.page-loaded")
+    _async_hide_locator = (By.CSS_SELECTOR, ".async-hide")
+    _login_locator = (By.CSS_SELECTOR, '[class="pardotTrackClick"]')
+    _user_nav_locator = (By.CSS_SELECTOR, '[class*="login-menu"]')
+    _logout_locator = (By.CSS_SELECTOR, "[href*=logout]")
+    _mobile_user_nav_locator = (By.CSS_SELECTOR, '[aria-label="Toggle Meta Navigation Menu"]')
+    _close_locator = (By.CSS_SELECTOR, '[class="close"]')
+    _mobile_user_nav_loaded_locator = (By.CSS_SELECTOR, '[aria-expanded="true"]')
+    _view_online_desktop_locator = (
+        By.XPATH,
+        "//div[@class='bigger-view']//span[text()='View online']/..",
+    )
+    _view_online_mobile_locator = (
+        By.XPATH,
+        "//div[@class='phone-view']//span[text()='View online']/..",
+    )
 
     @property
     def loaded(self):
         """Return True when the page-loaded class is added to the body tag."""
-        return (self.find_element(*self._root_locator)).is_displayed()
+        return (self.find_element(*self._root_locator)).is_displayed() and not self.find_elements(
+            *self._async_hide_locator
+        )
+
+    @property
+    def login(self):
+        return self.find_element(*self._login_locator)
+
+    @property
+    def user_nav(self):
+        return self.find_element(*self._user_nav_locator)
+
+    def mobile_user_nav_loaded(self):
+        return self.find_element(*self._mobile_user_nav_loaded_locator).is_displayed()
+
+    @property
+    def mobile_user_nav(self):
+        return self.find_element(*self._mobile_user_nav_locator)
+
+    @property
+    def user_is_logged_in(self):
+        if self.is_desktop:
+            if self.is_element_present(*self._user_nav_locator):
+                return True
+        elif self.is_mobile:
+            self.mobile_user_nav.click()
+            if self.is_element_present(*self._mobile_user_nav_locator):
+                self.mobile_user_nav.click()
+                self.wait.until(
+                    expected.invisibility_of_element_located(self._mobile_user_nav_loaded_locator)
+                )
+                return True
+
+    @property
+    def logout(self):
+        return self.find_element(*self._logout_locator)
+
+    @property
+    def view_online(self):
+        if self.is_desktop:
+            return self.find_element(*self._view_online_desktop_locator)
+        elif self.is_mobile:
+            return self.find_element(*self._view_online_mobile_locator)
+
+    def click_login(self):
+        if self.is_desktop:
+            self.login.click()
+        elif self.is_mobile:
+            self.click_mobile_user_nav()
+            self.login.click()
```

Comment:
Is there a third state possible? (ie not desktop nor mobile) If yes, ignore this :) otherwise the `elif self.is_mobile:` should probably be `else:` so each execution path is covered.